fix: Make supporting classes of AwsCredentials serializable

jmahonin opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-auth-library-java/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #1112 ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

davidrabinowitz commented:
This can fix https://github.com/GoogleCloudDataproc/spark-bigquery-connector/issues/845 as well. @jmahonin Do you need help with it?

jmahonin commented:
> This can fix https://github.com/GoogleCloudDataproc/spark-bigquery-connector/issues/845 as well. @jmahonin Do you need help with it?

Just a review at this point!

I can confirm this PR fixes the same serialization issue I observed with Flink. I'm not in a position to test against Spark, but based on past experience, I likewise suspect this will fix it there as well.

davidrabinowitz commented:
@jmahonin That's fine, I'll test it once released. Can you please rebase the PR so it can be merged?

jmahonin commented:
Will do, expect a ping on Monday

jmahonin commented:
@davidrabinowitz Rebased

TimurSadykov reviewed (changes requested):
Please add serialize unit tests for affected classes. You can leverage existing helper for it by extending the BaseSerializationTest

jmahonin commented:
Good call @TimurSadykov , added the tests and discovered `ExternalAccountCredentials.ServiceAccountImpersonationOptions` was not serializable. Added that as well.

There was already a serialization test on ImpersonatedCredentials, though they all use a mock transport factory so there isn't coverage on instantiating the factory in `readObject` (which is true of the existing serialization tests as well).

I didn't add a specific test for the deserialization of `SystemEnvironmentProvider`, it is similarly stubbed out with a `TestEnvironmentProvider` in the tests.

lsirac commented on oauth2_http/javatests/com/google/auth/oauth2/ExternalAccountCredentialsTest.java:
```diff
@@ -954,6 +956,37 @@ public void getRequestMetadata_withQuotaProjectId() throws IOException {
   }
 
+  @Test
+  public void serialize() throws IOException, ClassNotFoundException {
```
add similar tests in sub classes

jmahonin replied:
Will aim to have this updated tomorrow.

jmahonin commented:
I've updated the unit tests to ensure subclasses of ExternalAccountCredentials have similar `serialize()` tests.

However, the `PluggableAuthCredentials` uses a functional interface `PluggableAuthHandler`, which aren't easily serializable. Given the intent appears to be retrieving tokens using scripts / executables, it's not clear to me if this credential source should be serializable at all. Please advise.

TimurSadykov reviewed (approved):
LGTM

lsirac commented on oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:
```diff
@@ -66,8 +67,12 @@
  */
 public abstract class ExternalAccountCredentials extends GoogleCredentials {
 
+  private static final long serialVersionUID = 8049126194174465023L;
```
I think the subclasses should have their own private serialVersionUID.

jmahonin commented:
This fell off my radar for a bit. I see there's an approval and a request for changes re: subclasses having their own serialVersionUIDs. I'm happy to make the change, but want to confirm first given the PR's been approved.

lsirac commented:
[jmahonin](https://github.com/jmahonin) please make the change and we will merge the PR, LG otherwise

jmahonin commented:
> [jmahonin](https://github.com/jmahonin) please make the change and we will merge the PR, LG otherwise

This is done. I didn't include a serialVersionUID for `PluggableAuthCredentials` for the same reason described above here
https://github.com/googleapis/google-auth-library-java/pull/1113#issuecomment-1398489043
